### PR TITLE
Major Performance Improvements + Code Modernization via ES Module Support in Bolt

### PR DESCRIPTION
**Pull request by sghoweri**
## Jira
- http://vjira2:8080/browse/BDS-603
- http://vjira2:8080/browse/BDS-1828
- http://vjira2:8080/browse/BDS-1785
- http://vjira2:8080/browse/BDS-1834

<h1 id="related-prs">Related PRs Required To Ship</h1>

- [x] [Upgrade Webpack + Browserslist / Babel configs](https://github.com/boltdesignsystem/bolt/pull/1584)
- [x] [Internal component updates](https://github.com/boltdesignsystem/bolt/pull/1585)
- [x] [Consolidated / simplified polyfill loader updates](https://github.com/boltdesignsystem/bolt/pull/1587)
- [x] [Component Migration from Function-based to Class-based syntax](https://github.com/boltdesignsystem/bolt/pull/1591)
- [x] [Bolt Core / BoltElement Updates](https://github.com/boltdesignsystem/bolt/pull/1586)
- [x] [Link migration to LitElement](https://github.com/boltdesignsystem/bolt/pull/1588); related to getting the new decorator syntax wired up

# Related PRs Not Technically Required (Nice To Have)
- [x] [Device viewer CSS fix](https://github.com/boltdesignsystem/bolt/pull/1583)
- [x] [Pattern Lab UIKit fixes](https://github.com/boltdesignsystem/bolt/pull/1590) encountered while doing browser testing
- [x] [Lighthouse CI integration work to help us monitor the impact of incoming work on web performance and accessibility](https://github.com/boltdesignsystem/bolt/pull/1589)

## Summary
In a nutshell, this PR _partially_ migrates all of our existing components and build process over to optionally support delivering ES Modules (ES2015+ JavaScript) in supporting browsers when the new `esModules` config option is enabled in the build tools.

This necessary work preps us for further updates in the foreseeable future and helps inch us closer to several overarching goals, without having to do it all at once. These goals include:
- Improved code splitting and dead code removal
- Future migration to Rollup
- Ability to deliver all of Bolt's code via CDN (even if it's just for Codepen support, etc)
- Simpler polyfill strategy
- Overall faster load and render times**

**Be sure to check out the extensive web performance testing comparisons in the "Performance Testing Results" panel down below!

## Details

This work can be broken down into 3 general categories:
- The Build Tool and related config updates (Babel, Browserslist configs) — which can't get used if the code isn't updated to work with these under the hood updates 
- The actual component and supporting code updates (the vast majority of what had to change) — which can't get used if the build process isn't updated in the process
 - And the addition of Lighthouse CI to our Travis CI process to help with quantifying the initial performance improvements we'll be seeing with this work in addition to helping us monitor ongoing performance and accessibility work moving forward + flag potential issues.

C644CCA0-6006-4C68-8F07-595503CBC820

## Performance Impact

TLDR: While the work here won't necessarily reduce the total file size downloaded (albeit, this work gets us to that point), the actual JavaScript execution times + overhead when booting up and running components should be significantly reduced in modern browsers with minimal impact in IE 11.

> Note: these metrics were captured with the new modern JS bundle <script> tag not getting `async` loaded which if updated after further testing, could even further improve these numbers

#### JavaScript Main Thread in Chrome - Before vs After
image

#### JavaScript Main Thread in Mobile Chrome - Before vs After
image

#### Mobile 4G Start Render / Visually Complete Improvements
image
image

#### IE 11 Test Results (only ~0.5s slower on some of the heaviest pages)
image

<details>
  <summary>Performance Testing Results ⚡️</summary>

### Chrome, Cable (Before vs After) - about the same (~0.1s difference)
https://www.webpagetest.org/video/compare.php?tests=191204_RD_0f38729dd476283deae26ea2866cc0a6,191204_60_32c0c6c5c2c4b66862bff02bde35bd7e

### Mobile 3G Slow (Before vs After)
https://www.webpagetest.org/video/compare.php?tests=191204_Y1_0bbddb8a82f664157b8905aedba30676%2C191204_X6_1bc9919a593e64693e92c3ec5b9c6548&thumbSize=200&ival=100&end=full

- Icons show up ~0.7s faster
- Main image + Bands boot up ~1s faster
- Fully loaded is about the same (0.1s difference)

<hr>

#### Desktop (Before vs After)
https://www.webpagetest.org/video/compare.php?tests=191204_YA_027a200904154d80f9a22ee4529734b3,191204_NV_e79390218c5efa283a8bb7b9e287ab2c

image

#### Desktop (Before vs After), Take 2
https://www.webpagetest.org/video/compare.php?tests=191204_YA_027a200904154d80f9a22ee4529734b3%2C191204_NV_e79390218c5efa283a8bb7b9e287ab2c&thumbSize=200&ival=100&end=visual

image
> Despite a start render of 0.3s later (like the first test of this run on Desktop) visually complete in 0.8s sooner

#### Mobile 3G Slow (Before vs After)
https://www.webpagetest.org/video/compare.php?tests=191204_33_4b25994d88ff692e0889d9c863c82de0,191204_Y2_3a21f8cb8c8a6d446b206cd652f4f7d7
image
- 7.5s faster meaningful start render!
- Finishes loading ~19s faster!

#### Mobile 3G Slow (Before vs After), Take 2
https://www.webpagetest.org/video/compare.php?tests=191204_S0_622b6b7dc4d01d611ae93285d37bb8a4,191204_4X_236ee4098c9d9875dcefff58ea31f072

- 14.5s faster start render
- Finished loading 28s faster!

#### Mobile 3G Slow (Before vs After), Take 3
https://www.webpagetest.org/video/compare.php?tests=191204_S0_622b6b7dc4d01d611ae93285d37bb8a4,191204_4X_236ee4098c9d9875dcefff58ea31f072
image
- ~7.5s faster start render
- Visually complete ~19s faster!
- Fully loaded ~29s faster!

#### Mobile 3G Slow (Before vs After), Take 4
https://www.webpagetest.org/video/compare.php?tests=191204_33_4b25994d88ff692e0889d9c863c82de0%2C191204_Y2_3a21f8cb8c8a6d446b206cd652f4f7d7&thumbSize=200&ival=1000&end=visual
image

- ~7s faster start render (~15s down to 8s)
- Visually complete ~19s faster (39s down to 20s)
- Fully loaded ~29s faster! (53s down to 24s)

#### Mobile LTE (Before vs After)
https://www.webpagetest.org/video/compare.php?tests=191204_GC_f53955c46ce6d988f0cd3dbd67218169%2C191204_KZ_d05b0ca6384ef5d960406e4d778e214b&thumbSize=200&ival=500&end=visual

image

- Similar first paint
- Icons paint ~1s faster
- Fully loads in 3.4s faster!
- Only 2KB “size” difference between the two: savings are all related to execution time vs file size

<hr>

#### Mobile 3G Slow

https://www.webpagetest.org/video/compare.php?tests=191204_1N_98edc69f642afca33eb8d70723c11c69%2C191204_EY_e3d982d22bdbb63008ed98fc46d9a801&thumbSize=200&ival=1000&end=visual

image

- Similar start times
- Hero image shows up 1s faster (19.8s vs 18.8s)
- Visually complete 1s faster (20.2s vs 19.2s)

#### Mobile 4G
https://www.webpagetest.org/video/compare.php?tests=191204_6C_b8f4ac4696cd5ad5c297047a13a96209%2C191204_MM_cf4d8241e8a6bb42b080ef89b9d98999&thumbSize=200&ival=100&end=visual

image

- Start render times similar (~0.2s difference)
- Hero image renders ~0.7s faster (7.4s vs 6.7s)
- Visually complete 0.6s faster (7.5s down to 6.9s)

<hr>

#### Desktop (Chrome)
https://www.webpagetest.org/video/compare.php?tests=191204_PE_aeebea5aa50b0f97236dcc1a25930031%2C191204_ZC_2f20110c37e9d2ac349062c55616adfd&thumbSize=200&ival=500&end=full

image

- Buttons render ~0.7s faster (3.3s down to 2.6s)
- Fully loaded in ~0.6s faster (4.1s down to 3.5s)

#### Mobile 4G

https://www.webpagetest.org/video/compare.php?tests=191204_B3_f8f080f27e566c3e4488b7327c064bf4%2C191204_W3_952ebdbed21244d0b87e7f4d59f17321&thumbSize=200&ival=100&end=full

image

- Fully loads ~1s faster (10.4s down to 9.5s)
- Buttons show up 1.2s faster (8.5s vs 7.3s)
- Micro Journey logo shows up 0.9s faster (8.7s down to 7.8s)

</details>

## How to test
See linked PRs for guidelines on the different moving parts involved + things to keep an eye on.

### CODE REVIEW

**Comment by sghoweri** [comment]
Thanks @adamszalapski!
